File: frontend/src/tipos/suscripcion.ts
import type { Nullable } from './api';

export type EstadoSuscripcion = 'ACTIVA' | 'TRIAL' | 'CANCELADA' | 'SUSPENDIDA' | 'VENCIDA';
export type CicloSuscripcion = 'MENSUAL' | 'ANUAL';

export interface Suscripcion {
  id: string;
  tenant_id: string;
  plan_id: string;
  estado: EstadoSuscripcion;
  ciclo: CicloSuscripcion;
  fecha_inicio: string;
  fecha_fin: Nullable<string>;
  fecha_proximo_cobro: Nullable<string>;
  es_trial: boolean;
  dias_trial: number;
  fecha_fin_trial: Nullable<string>;
  override_max_sedes: Nullable<number>;
  override_max_usuarios: Nullable<number>;
  override_max_reclamos: Nullable<number>;
  override_max_chatbots: Nullable<number>;
  override_max_canales_whatsapp: Nullable<number>;
  override_max_storage_mb: Nullable<number>;
  referencia_pago: Nullable<string>;
  metodo_pago: Nullable<string>;
  activado_por: Nullable<string>;
  activado_por_usuario_id: Nullable<string>;
  activado_por_usuario_nombre: Nullable<string>;
  notas: Nullable<string>;
  fecha_creacion: string;
  fecha_actualizacion: string;
}

// Respuesta de GET /suscripcion (suscripción + plan)
export interface SuscripcionConPlan {
  suscripcion: Suscripcion;
  plan: import('./plan').Plan;
}

// Respuesta de GET /suscripcion/uso
export interface UsoTenant {
  tenant_id: string;

  // Plan y suscripción
  plan_id: string;
  plan_codigo: string;
  plan_nombre: string;
  suscripcion_id: string;
  suscripcion_estado: string;
  suscripcion_ciclo: string;
  es_trial: boolean;

  // Límites efectivos (-1 = ilimitado)
  limite_sedes: number;
  limite_usuarios: number;
  limite_reclamos_mes: number;
  limite_chatbots: number;
  limite_canales_whatsapp: number;
  limite_storage_mb: number;

  // Funcionalidades
  permite_chatbot: boolean;
  permite_whatsapp: boolean;
  permite_email: boolean;
  permite_reportes_pdf: boolean;
  permite_exportar_excel: boolean;
  permite_api: boolean;
  permite_marca_blanca: boolean;
  permite_multi_idioma: boolean;
  permite_asistente_ia: boolean;
  permite_atencion_vivo: boolean;

  // Uso actual
  uso_sedes: number;
  uso_usuarios: number;
  uso_reclamos_mes: number;
  uso_chatbots: number;
  uso_canales_whatsapp: number;
}

export interface CambiarPlanRequest {
  plan_codigo: string;
  ciclo: CicloSuscripcion;
  referencia_pago?: string;
  metodo_pago?: string;
  notas?: string;
}

// Helper: porcentaje de uso
export function porcentajeUso(uso: number, limite: number): number {
  if (limite === -1) return -1; // ilimitado
  if (limite === 0) return 100;
  return Math.round((uso / limite) * 100);
}

File: backend/internal/model/suscripcion.go
package model

import (
	"time"

	"github.com/google/uuid"
)

// Suscripcion vincula un tenant con un plan.
// Un tenant tiene UNA suscripción activa a la vez.
type Suscripcion struct {
	TenantModel

	PlanID uuid.UUID `json:"plan_id" db:"plan_id"`
	Estado string    `json:"estado" db:"estado"`
	Ciclo  string    `json:"ciclo" db:"ciclo"`

	FechaInicio       time.Time `json:"fecha_inicio" db:"fecha_inicio"`
	FechaFin          NullTime  `json:"fecha_fin" db:"fecha_fin"`
	FechaProximoCobro NullTime  `json:"fecha_proximo_cobro" db:"fecha_proximo_cobro"`

	// ── Trial ──
	EsTrial       bool     `json:"es_trial" db:"es_trial"`
	DiasTrial     int      `json:"dias_trial" db:"dias_trial"`
	FechaFinTrial NullTime `json:"fecha_fin_trial" db:"fecha_fin_trial"`

	// ── Overrides (NULL = usa límite del plan) ──
	OverrideMaxSedes            NullInt64 `json:"override_max_sedes" db:"override_max_sedes"`
	OverrideMaxUsuarios         NullInt64 `json:"override_max_usuarios" db:"override_max_usuarios"`
	OverrideMaxReclamos         NullInt64 `json:"override_max_reclamos" db:"override_max_reclamos"`
	OverrideMaxChatbots         NullInt64 `json:"override_max_chatbots" db:"override_max_chatbots"`
	OverrideMaxCanalesWhatsApp  NullInt64 `json:"override_max_canales_whatsapp" db:"override_max_canales_whatsapp"`
	OverrideMaxStorageMB        NullInt64 `json:"override_max_storage_mb" db:"override_max_storage_mb"`

	// ── Pago ──
	ReferenciaPago NullString `json:"referencia_pago" db:"referencia_pago"`
	MetodoPago     NullString `json:"metodo_pago" db:"metodo_pago"`

	ActivadoPor          NullString `json:"activado_por" db:"activado_por"`
	ActivadoPorUsuarioID NullUUID   `json:"activado_por_usuario_id" db:"activado_por_usuario_id"`
	Notas                NullString `json:"notas" db:"notas"`

	// Campo computed (JOIN) — no se persiste directamente
	ActivadoPorUsuarioNombre NullString `json:"activado_por_usuario_nombre,omitempty" db:"-"`

	FechaCreacion      time.Time `json:"fecha_creacion" db:"fecha_creacion"`
	FechaActualizacion time.Time `json:"fecha_actualizacion" db:"fecha_actualizacion"`
}

// Estados de suscripción.
const (
	SuscripcionActiva    = "ACTIVA"
	SuscripcionTrial     = "TRIAL"
	SuscripcionSuspendida = "SUSPENDIDA"
	SuscripcionCancelada = "CANCELADA"
	SuscripcionVencida   = "VENCIDA"
)

// Ciclos de facturación.
const (
	CicloMensual = "MENSUAL"
	CicloAnual   = "ANUAL"
)

// Orígenes de activación.
const (
	ActivadoPorOnboarding = "ONBOARDING"
	ActivadoPorUpgrade    = "UPGRADE"
	ActivadoPorAdmin      = "ADMIN_MANUAL"
	ActivadoPorRenovacion = "RENOVACION"
)

// Métodos de pago soportados.
const (
	MetodoPagoTarjeta      = "TARJETA"
	MetodoPagoTransferencia = "TRANSFERENCIA"
	MetodoPagoYape         = "YAPE"
	MetodoPagoPlin         = "PLIN"
)

// EstaActiva indica si la suscripción permite operar.
func (s *Suscripcion) EstaActiva() bool {
	return s.Estado == SuscripcionActiva || s.Estado == SuscripcionTrial
}

// EstaVencida verifica si el trial o la suscripción venció.
func (s *Suscripcion) EstaVencida() bool {
	if s.EsTrial && s.FechaFinTrial.Valid {
		return time.Now().After(s.FechaFinTrial.Time)
	}
	if s.FechaFin.Valid {
		return time.Now().After(s.FechaFin.Time)
	}
	return false
}
